src/lib.rs

```diff
@@ -308,55 +308,154 @@ impl Reflection {
     ) -> Result<u32, ReflectError> {
         debug_assert!(struct_instruction.class.opcode == spirv::Op::TypeStruct);
 
-        // if there are less then two members there is no offset to use
+        // if there are less then two members there is no offset to use, early out
         if struct_instruction.operands.len() < 2 {
             return Ok(0);
         }
 
-        let result_id = match struct_instruction.result_id {
-            Some(id) => id,
-            None => return Err(ReflectError::MissingResultId(struct_instruction.clone())),
-        };
+        let result_id = struct_instruction
+            .result_id
+            .ok_or_else(|| ReflectError::MissingResultId(struct_instruction.clone()))?;
 
-        let variable = reflect
+        let variable_id = reflect
             .debugs
             .iter()
             .filter(|i| i.class.opcode == spirv::Op::Name)
             .find_map(|i| {
-                if get_operand_at!(i, Operand::IdRef, 0).ok()? == result_id {
-                    Some(i)
+                let id = get_operand_at!(i, Operand::IdRef, 0).ok()?;
+                if id == result_id {
+                    Some(id)
                 } else {
                     None
                 }
-            });
-
-        let variable = match variable {
-            Some(i) => i,
-            None => return Err(ReflectError::UnknownStruct(struct_instruction.clone())),
-        };
+            })
+            .ok_or_else(|| ReflectError::UnknownStruct(struct_instruction.clone()))?;
 
-        let idref = get_operand_at!(variable, Operand::IdRef, 0)?;
-        let annotations = Self::find_annotations_for_id(&reflect.annotations, idref)?;
+        let annotations = Self::find_annotations_for_id(&reflect.annotations, variable_id)?;
         let member_decorates = annotations
             .iter()
             .filter(|i| i.class.opcode == spirv::Op::MemberDecorate)
             .collect::<Vec<_>>();
 
-        // return the value from the last offset member decoration
-        for i in member_decorates.iter().rev() {
-            let decoration = get_operand_at!(**i, Operand::Decoration, 2)?;
-            if let spirv::Decoration::Offset = decoration {
-                return Ok(get_operand_at!(**i, Operand::LiteralInt32, 3)?);
-            }
-        }
-
-        Ok(0)
+        // return the highest offset value
+        Ok(*member_decorates
+            .iter()
+            .filter_map(|i| match get_operand_at!(**i, Operand::Decoration, 2) {
+                Ok(decoration) if decoration == spirv::Decoration::Offset => {
+                    match get_operand_at!(**i, Operand::LiteralInt32, 3) {
+                        Ok(val) => Some(Ok(val)),
+                        Err(err) => Some(Err(err)),
+                    }
+                }
+                Err(err) => Some(Err(err)),
+                _ => None,
+            })
+            .collect::<Result<Vec<_>>>()?
+            .iter()
```

Comment:
If you make this `.into_iter()` the `Vec` is consumed and the result of `.max()` returns `u32` instead of `&u32`. In turn you  can remove reference handling in `Ok(*member_decorates` and `.unwrap_or(&0)`.